compiler/rustc_codegen_llvm/src/coverageinfo/map_data.rs

```diff
@@ -12,28 +13,21 @@ pub struct Expression {
     lhs: CovTerm,
     op: Op,
     rhs: CovTerm,
-    code_regions: Vec<CodeRegion>,
 }
 
-/// Collects all of the coverage regions associated with (a) injected counters, (b) counter
-/// expressions (additions or subtraction), and (c) unreachable regions (always counted as zero),
-/// for a given Function. This struct also stores the `function_source_hash`,
-/// computed during instrumentation, and forwarded with counters.
-///
-/// Note, it may be important to understand LLVM's definitions of `unreachable` regions versus "gap
-/// regions" (or "gap areas"). A gap region is a code region within a counted region (either counter
-/// or expression), but the line or lines in the gap region are not executable (such as lines with
-/// only whitespace or comments). According to LLVM Code Coverage Mapping documentation, "A count
-/// for a gap area is only used as the line execution count if there are no other regions on a
-/// line."
+/// Holds all of the coverage mapping data associated with a function instance,
+/// collected during traversal of `Coverage` statements in the function's MIR.
 #[derive(Debug)]
 pub struct FunctionCoverage<'tcx> {
     /// Coverage info that was attached to this function by the instrumentor.
     function_coverage_info: &'tcx FunctionCoverageInfo,
     is_used: bool,
-    counters: IndexVec<CounterId, Option<Vec<CodeRegion>>>,
+
+    /// Tracks which counters have been seen, to avoid duplicate mappings
+    /// that might be introduced by MIR inlining.
```

Comment:
Before inlining, counter IDs (and expression IDs) are local to their enclosing function.

After inlining, we use source scopes to identify which function a coverage statement came from. This is handled in <https://github.com/rust-lang/rust/blob/master/compiler/rustc_codegen_ssa/src/mir/coverageinfo.rs>.

So when `FunctionCoverage` sees a coverage statement, it doesn't have to worry about inlining. All of the statements it sees either weren't inlined (and thus belong to its `instance`), or were inlined *from* the function that eventually because its `instance`. So it can assume that the counter/expression IDs it sees are ones that make sense for this instance.

Reply:
(The fact that counter/expression IDs are function-local might not be explicitly documented anywhere; it's one of those things that became so obvious to me that I didn't think to write it down.)

Reply:
And yeah, if a `CounterIncrement` or `ExpressionUsed` is inlined into another function, the original instance may see multiple copies of its own counter/expression IDs. (The function being inlined into won’t see those IDs.)